dbms/src/Databases/DatabaseOrdinary.cpp

```diff
@@ -97,57 +109,78 @@ DatabaseOrdinary::DatabaseOrdinary(String name_, const String & metadata_path_,
+            if (ast)
+            {
+                auto * create_query = ast->as<ASTCreateQuery>();
+                file_names[file_name] = ast;
+                total_dictionaries += create_query->is_dictionary;
+            }
+        }
+        catch (const Exception & e)
+        {
+            throw Exception(
+                "Cannot parse definition from metadata file " + full_path + ", error: " + e.displayText() + ", stack trace:\n"
+                    + e.getStackTrace().toString(), ErrorCodes::CANNOT_PARSE_TEXT);
+        }
+
+    });
+
+    size_t total_tables = file_names.size() - total_dictionaries;
 
-    const size_t total_tables = file_names.size();
-    LOG_INFO(log, "Total " << total_tables << " tables.");
+    LOG_INFO(log, "Total " << total_tables << " tables and " << total_dictionaries << " dictionaries.");
 
     AtomicStopwatch watch;
-    std::atomic<size_t> tables_processed {0};
+    std::atomic<size_t> tables_processed{0};
+    std::atomic<size_t> dictionaries_processed{0};
 
-    auto loadOneTable = [&](const String & table)
+    auto loadOneObject = [&](const ASTCreateQuery & query)
     {
-        loadTable(context, getMetadataPath(), *this, getDatabaseName(), getDataPath(), table, has_force_restore_data_flag);
+        loadObject(context, query, *this, getDataPath(), getDatabaseName(), has_force_restore_data_flag);
 
         /// Messages, so that it's not boring to wait for the server to load for a long time.
-        if (++tables_processed % PRINT_MESSAGE_EACH_N_TABLES == 0
-            || watch.compareAndRestart(PRINT_MESSAGE_EACH_N_SECONDS))
-        {
-            LOG_INFO(log, std::fixed << std::setprecision(2) << tables_processed * 100.0 / total_tables << "%");
-            watch.restart();
-        }
+        if (query.is_dictionary)
+            logAboutProgress(log, ++dictionaries_processed, total_dictionaries, watch);
+        else
+            logAboutProgress(log, ++tables_processed, total_tables, watch);
     };
 
     ThreadPool pool(SettingMaxThreads().getAutoValue());
 
-    for (const auto & file_name : file_names)
+    for (const auto & name_with_query : file_names)
     {
-        pool.scheduleOrThrowOnError([&]() { loadOneTable(file_name); });
+        pool.scheduleOrThrowOnError([&]() { loadOneObject(name_with_query.second->as<const ASTCreateQuery &>()); });
     }
 
     pool.wait();
 
     /// After all tables was basically initialized, startup them.
     startupTables(pool);
+
+    /// Add database as repository
+    auto dictionaries_repository = std::make_unique<ExternalLoaderDatabaseConfigRepository>(shared_from_this(), context);
+    auto & external_loader = context.getExternalDictionariesLoader();
+    external_loader.addConfigRepository(getDatabaseName(), std::move(dictionaries_repository));
+    bool lazy_load = context.getConfigRef().getBool("dictionaries_lazy_load", true);
+    external_loader.reload(!lazy_load);
```

Comment:
So after this change all dictionaries will be reloaded, regardless the `lifetime`?
And this can be not the desired behavior if the dictionary is big enough

Reply:
@alesapin ?